artifact/v3/version/version.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <string>
+#include <vector>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/json.hpp>
+
+#include <artifact/v3/version/version.hpp>
+
+namespace artifact {
+namespace v3 {
+namespace version {
+
+namespace io = mender::common::io;
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace json = mender::common::json;
+
+const int supported_parser_version {3};
+const string supported_parser_format {"mender"};
+
+const ErrorCategoryClass ErrorCategory {};
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "VersionParserErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	ParseError:
+		return "Parse error";
+	VersionError:
+		return "Wrong Artifact version";
+	FormatError:
+		return "Wrong Artifact format";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+expected::Expected<Version, error::Error> Parse(io::Reader &reader) {
+	// Read in all the bytes into a tmp buffer
+	std::vector<uint8_t> buf(1024);
+	io::ByteWriter bw {buf};
+
+	auto err = io::Copy(bw, reader);
```

Comment:
At least from what I'm used to, `Copy` is nice ,because it guarantees to read to the end. A single `Read` call can theoretically do short reads, and may need to be called again.

Reply:
But the data is shuffled around (into an internal buffer in `Copy` and then into the `ByteWriter` above) and you can easily get an error if `buf` above is not big enough. Like I said above, to me it would make much more sense to have a function/class/... to read everything into a buffer and then use that buffer.

Reply:
This is true in the general case. But the advantage here is that we control the format, and the `version` header is never going to be bigger than what is here available.

Reply:
> But the data is shuffled around (into an internal buffer in Copy and then into the ByteWriter above) and you can easily get an error if buf above is not big enough. Like I said above, to me it would make much more sense to have a function/class/... to read everything into a buffer and then use that buffer.

Isn't this what I am doing with the `byte-buffer` above already?

Reply:
Assuming it's at most 1024 bytes. But let's not ride a dead horse here anymore. I'll try to come up with something for this myself and then we can have a discussion there.

Reply:
> So there are 3 copies of the same thing, all with read-only access.

It's 4 actually. `Copy` has its own internal buffer too.